rfcs/20200929-keras-mixed-precision.md

```diff
@@ -0,0 +1,1178 @@
+
+`increment_period` defaults to 2000. There is a tradeoff between making it too low or high. The advantage of a low `increment_period` is it allows the loss scale to rise faster, which is good if the optimal loss scale tends to increase quickly during training. The disadvantage of a low `increment_period` is that it skips more steps. The default of 2000 is arbitrary. We have not run experiments on the optimal `increment_period`. but setting it to 2000 works well in practice. The default causes 0.05% of the steps to be skipped on average.
+
+`multiplier` defaults to 2, because this causes the loss scale to always be a power of 2 (assuming the `initial_loss_scale` is also a power of 2). Scaling loss and gradients by powers of 2 causes no loss in precision, unlike scaling by an arbitrary factor.
+
+## LossScaleOptimizer Overview
+
+LossScaleOptimizer wraps an existing optimizer and implements loss scaling. It will be exposed as `tf.keras.mixed_precision.LossScaleOptimizer` and is currently exposed in TensorFlow 2.3 as [`tf.keras.mixed_precision.experimental.LossScaleOptimizer`](https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/experimental/LossScaleOptimizer).
+
+The LossScaleOptimizer constructor takes in three arguments:
+
+* `optimizer`: The optimizer to wrap.
+* `loss_scale`, the LossScale to use, defaulting to `DynamicLossScale()`. Like any function taking in a loss scale, a Python int or float can be passed, which is equivalent to passing the corresponding FixedLossScale. The string "dynamic" can also be passed, which is equivalent to passing `DynamicLossScale()`.
+* `aggregate_in_float16` (optional): Boolean defaulting to True. If True, `apply_gradients` will all-reduce in float16 instead of the original dtype of the gradients. This will likely be initially set to False in TensorFlow 2.4 and switched to True in a future release. See the [All-reducing in float16](#all-reducing-in-float16)" section for details.
+
+LossScaleOptimizer subclasses `tf.keras.optimizers.Optimizer`. For the most part, LossScaleOptimizer simply overrides the base Optimizer methods to delegate to the wrapped optimizer, but adds loss scaling to certain methods.
+
+There are four steps that LossScaleOptimizer must perform in order to implement loss scaling:
+
+1. Multiply the loss by the loss scale.
+2. Divide the gradients by the loss scale.
+3. For a DynamicLossScale, update the loss scale. This means increasing or decreasing the loss scale and updating `num_good_steps` in accordance with the dynamic loss scaling algorithm.
+4. For a DynamicLossScale, skip applying gradients if they are not finite. Gradients are not finite if they have an Inf, -Inf, or NaN value. If any gradient of any replica has a nonfinite value, all gradients across all replicas are skipped for that step. For a FixedLossScale, gradients are unconditionally applied, just like when loss scaling is not used.
```

Comment:
What is the benefit of skipping update steps instead of recomputing the gradients with a scaled down loss until the gradients become finite (or until fixed number of retries is reached)? Intuitively this would enable slightly higher performance as the forward pass could be reused, or am I missing something?

Reply:
This is a good idea, but the issue is that we cannot run the loop if the user calls `apply_gradients` instead of `minimize` because `apply_gradients` does not compute the gradients. For example:

```
vars = ...
opt = tf.keras.mixed_precision.LossScaleOptimizer(...)
with tf.GradientTape() as tape:
  loss = get_model_loss()
  grads = tape.gradient(loss, vars)
  opt.apply_gradients(zip(grads, vars))
```

Since the gradients are computed outside the LossScaleOptimizer, LossScaleOptimizer has no way of repeatedly recomputing the gradients.

However, we can do this in `minimize`. I added a paragraph suggesting we should add an option to do this in the future (starting with "Instead of skipping steps when there are NaNs"). The reason this won't be done initially is that `minimize` is rarely used, but I anticipate it will be used more frequently due to #234 allowing users to pass a tensor to `minimize` once TF 2.4 is released.

I think the performance gain will be negligible, due to steps only being skipped 1/2000 of the time on average, but the behavior will be more intuitive. The skipping behavior is confusing because for the first few steps, the variables are not actually updated since the loss scale starts out so high.

Reply:
Yes, sorry I should've been clearer. I was mainly suggesting this for the use in `keras.Model.fit()` which indeed calls [`.minimize()`](https://github.com/tensorflow/tensorflow/blob/dc843fad65fb4eb1b173a7577d0d932b2af94951/tensorflow/python/keras/engine/training.py#L743). But since for custom training loops the skipping behaviour still needs to be supported anyway, I agree it makes sense to keep the skipping behaviour for now.